File: backend/config/dotenv.ts
// Reads a .env file into the process environment.
//
// The dashboard is started by hand on a venue machine, often by double-clicking
// or by a one-line `npm start` — there is no launcher (mise, direnv, dotenv-cli)
// to populate the environment first, and a missing AWS_REGION shows up much
// later as an empty panel. So the process loads its own .env.
//
// Two rules keep this predictable:
//   * A variable already present in the real environment always wins. Exporting
//     a value, or `API_ADDR=… npm start`, overrides the file without editing it.
//   * A missing file is not an error. Every value here has a default or lives
//     in the settings table.
//
// Go's backend/internal/config/dotenv.go parses the same grammar; keep the two
// in step.

import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

/** One key/value pair as written in the file. */
export type EnvPairs = Record<string, string>;

/** Where a .env was found, and what it contributed. */
export interface DotenvResult {
  /** Absolute path of the file that was read, or null if none existed. */
  path: string | null;
  /** Keys taken from the file (those not already set in the environment). */
  applied: string[];
  /** Keys present in the file but skipped because the environment had them. */
  skipped: string[];
}

/**
 * Candidate paths, in order. The first that exists wins.
 *
 * ENV_FILE is the explicit escape hatch (two instances, two configs). The
 * working directory is the normal case. The directory holding this source tree
 * covers `npm start` run from somewhere else — a shortcut's "start in" field on
 * Windows is a common way to get a surprising cwd.
 */
function candidates(explicit?: string): string[] {
  const list: string[] = [];
  const fromEnv = (explicit ?? process.env.ENV_FILE ?? "").trim();
  if (fromEnv !== "") list.push(resolve(fromEnv));
  list.push(resolve(process.cwd(), ".env"));
  // backend/config/dotenv.ts -> repository root.
  list.push(resolve(dirname(fileURLToPath(import.meta.url)), "..", "..", ".env"));
  return [...new Set(list)];
}

/**
 * Parses .env text. Supports `KEY=value`, an optional `export ` prefix, `#`
 * comments, single-quoted (literal) and double-quoted (escapes) values, and
 * trailing comments after an unquoted value. Malformed lines are skipped rather
 * than thrown: a stray line in a config file must not stop the dashboard from
 * booting mid-exercise.
 */
export function parseEnv(text: string): EnvPairs {
  const out: EnvPairs = {};
  // A leading BOM would otherwise become part of the first key — Windows
  // editors add one, and the resulting "﻿AWS_REGION" silently reads empty.
  for (const raw of text.replace(/^﻿/, "").split(/\r?\n/)) {
    const line = raw.trim();
    if (line === "" || line.startsWith("#")) continue;

    const body = line.startsWith("export ") ? line.slice("export ".length).trim() : line;
    const eq = body.indexOf("=");
    if (eq <= 0) continue;

    const key = body.slice(0, eq).trim();
    if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(key)) continue;

    out[key] = parseValue(body.slice(eq + 1).trim());
  }
  return out;
}

function parseValue(raw: string): string {
  if (raw.startsWith("'") && raw.length > 1) {
    const end = raw.indexOf("'", 1);
    // Single quotes are literal: no escapes, no interpolation.
    return end < 0 ? raw.slice(1) : raw.slice(1, end);
  }
  if (raw.startsWith('"') && raw.length > 1) {
    const end = findClosingDouble(raw);
    const inner = end < 0 ? raw.slice(1) : raw.slice(1, end);
    return inner.replace(/\\(.)/g, (_, c: string) =>
      c === "n" ? "\n" : c === "r" ? "\r" : c === "t" ? "\t" : c,
    );
  }
  // Unquoted: a ` #` starts a trailing comment. A bare `#` inside a token (a
  // URL fragment, a password) is kept. A value that starts where a comment
  // would is an empty value with a note after it: `KEY=  # note`.
  if (raw.startsWith("#")) return "";
  const comment = raw.search(/\s#/);
  return (comment < 0 ? raw : raw.slice(0, comment)).trim();
}

/** Index of the closing double quote, skipping escaped ones. */
function findClosingDouble(raw: string): number {
  for (let i = 1; i < raw.length; i++) {
    if (raw[i] === "\\") {
      i++;
      continue;
    }
    if (raw[i] === '"') return i;
  }
  return -1;
}

/**
 * Loads the first .env found into process.env without overwriting anything the
 * environment already provides. Safe to call more than once (the launcher and
 * the backend both do, in different processes).
 */
export function loadDotenv(explicitPath?: string): DotenvResult {
  for (const path of candidates(explicitPath)) {
    let text: string;
    try {
      text = readFileSync(path, "utf8");
    } catch {
      continue; // Absent, or unreadable — try the next candidate.
    }

    const applied: string[] = [];
    const skipped: string[] = [];
    for (const [key, value] of Object.entries(parseEnv(text))) {
      // An empty value in .env means "unset" — the file ships with blank AWS
      // keys as documentation, and turning those into empty strings would
      // shadow the credentials the settings screen injects later.
      if (value === "") continue;
      if (process.env[key] !== undefined && process.env[key] !== "") {
        skipped.push(key);
        continue;
      }
      process.env[key] = value;
      applied.push(key);
    }
    return { path, applied, skipped };
  }
  return { path: null, applied: [], skipped: [] };
}

File: backend/internal/config/dotenv.go
package config

// Reads a .env file into the process environment.
//
// The dashboard is started by hand on a venue machine — often by
// double-clicking the release binary — so there is no launcher (mise, direnv,
// dotenv-cli) to populate the environment first, and a missing AWS_REGION shows
// up much later as an empty panel. The process therefore loads its own .env.
//
// Two rules keep this predictable:
//   - A variable already present in the real environment always wins. Exporting
//     a value overrides the file without editing it.
//   - A missing file is not an error. Every value here has a default or lives in
//     the settings table.
//
// backend/config/dotenv.ts in the Node port parses the same grammar; keep the
// two in step.

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// Dotenv reports what a load found: the file that was read (empty if none), the
// keys taken from it, and the keys it yielded to the environment on.
type Dotenv struct {
	Path    string
	Applied []string
	Skipped []string
}

var envKeyRe = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)

// dotenvCandidates lists the paths to try, in order; the first that reads wins.
//
// ENV_FILE is the explicit escape hatch (two instances, two configs). The
// working directory is the normal case. The directory holding the executable
// covers a release binary launched from Explorer, where the working directory
// is whatever the shortcut says and rarely where the .env sits.
func dotenvCandidates(explicit string) []string {
	var list []string
	add := func(p string) {
		if p == "" {
			return
		}
		abs, err := filepath.Abs(p)
		if err != nil {
			abs = p
		}
		for _, seen := range list {
			if seen == abs {
				return
			}
		}
		list = append(list, abs)
	}

	if explicit == "" {
		explicit = strings.TrimSpace(os.Getenv("ENV_FILE"))
	}
	add(explicit)
	add(".env")
	if exe, err := os.Executable(); err == nil {
		add(filepath.Join(filepath.Dir(exe), ".env"))
	}
	return list
}

// ParseEnv parses .env text. It supports KEY=value, an optional "export "
// prefix, # comments, single-quoted (literal) and double-quoted (escapes)
// values, and a trailing comment after an unquoted value. Malformed lines are
// skipped rather than reported: a stray line in a config file must not stop the
// dashboard from booting mid-exercise.
func ParseEnv(text string) map[string]string {
	out := map[string]string{}
	// A leading BOM would otherwise become part of the first key — Windows
	// editors add one, and the resulting "<BOM>AWS_REGION" silently reads empty.
	text = strings.TrimPrefix(text, "\ufeff")

	for _, raw := range strings.Split(strings.ReplaceAll(text, "\r\n", "\n"), "\n") {
		line := strings.TrimSpace(raw)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		body := line
		if rest, ok := strings.CutPrefix(body, "export "); ok {
			body = strings.TrimSpace(rest)
		}
		key, value, ok := strings.Cut(body, "=")
		if !ok {
			continue
		}
		key = strings.TrimSpace(key)
		if !envKeyRe.MatchString(key) {
			continue
		}
		out[key] = parseEnvValue(strings.TrimSpace(value))
	}
	return out
}

func parseEnvValue(raw string) string {
	if strings.HasPrefix(raw, "'") && len(raw) > 1 {
		// Single quotes are literal: no escapes, no interpolation.
		if end := strings.Index(raw[1:], "'"); end >= 0 {
			return raw[1 : 1+end]
		}
		return raw[1:]
	}
	if strings.HasPrefix(raw, `"`) && len(raw) > 1 {
		inner := raw[1:]
		if end := closingDouble(raw); end >= 0 {
			inner = raw[1:end]
		}
		return unescapeDouble(inner)
	}
	// Unquoted: whitespace then # starts a trailing comment. A bare # inside a
	// token (a URL fragment, a password) is kept. A value that starts where a
	// comment would is an empty value with a note after it: "KEY=  # note".
	if strings.HasPrefix(raw, "#") {
		return ""
	}
	for i := 1; i < len(raw); i++ {
		if raw[i] == '#' && (raw[i-1] == ' ' || raw[i-1] == '\t') {
			return strings.TrimSpace(raw[:i])
		}
	}
	return raw
}

// closingDouble is the index of the closing double quote, skipping escaped ones.
func closingDouble(raw string) int {
	for i := 1; i < len(raw); i++ {
		if raw[i] == '\\' {
			i++
			continue
		}
		if raw[i] == '"' {
			return i
		}
	}
	return -1
}

func unescapeDouble(s string) string {
	var b strings.Builder
	for i := 0; i < len(s); i++ {
		if s[i] != '\\' || i+1 >= len(s) {
			b.WriteByte(s[i])
			continue
		}
		i++
		switch s[i] {
		case 'n':
			b.WriteByte('\n')
		case 'r':
			b.WriteByte('\r')
		case 't':
			b.WriteByte('\t')
		default:
			b.WriteByte(s[i])
		}
	}
	return b.String()
}

// LoadDotenv folds the first .env found into the process environment without
// overwriting anything the environment already provides. Pass an empty path to
// use the normal search order.
func LoadDotenv(explicit string) Dotenv {
	for _, path := range dotenvCandidates(explicit) {
		data, err := os.ReadFile(path)
		if err != nil {
			continue // Absent, or unreadable — try the next candidate.
		}
		res := Dotenv{Path: path}
		for key, value := range ParseEnv(string(data)) {
			// An empty value in .env means "unset" — the file ships with blank
			// AWS keys as documentation, and turning those into empty strings
			// would shadow the credentials the settings screen injects later.
			if value == "" {
				continue
			}
			if existing := os.Getenv(key); existing != "" {
				res.Skipped = append(res.Skipped, key)
				continue
			}
			if err := os.Setenv(key, value); err != nil {
				continue
			}
			res.Applied = append(res.Applied, key)
		}
		return res
	}
	return Dotenv{}
}
